### PR TITLE
ascanrules: add example alerts to Format String Error

### PR DESCRIPTION
**Pull request by GRVial**
- Added example alerts for [Format String Error](https://www.zaproxy.org/docs/alerts/30002/)
- Added unit test for example alerts
- Added note to CHANGELOG

Issue Related: [zaproxy/zaproxy#6119](https://github.com/zaproxy/zaproxy/issues/6119)

### CODE REVIEW

**Comment by thc202** [comment]
The build is failing: https://github.com/zaproxy/zap-extensions/actions/runs/5083056477/jobs/9133572333?pr=4623#step:6:475

The commits can be squashed.

File: addOns/ascanrules/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## [54] - 2023-05-03
 ### Changed
 - Maintenance changes.
+- The Format String Error scan rule now includes example alert functionality for documentation generation purposes (Issue 6119).
```
Comment (thc202) [inline addOns/ascanrules/CHANGELOG.md:14]: This should be moved to unreleased section.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java [hunk 2]
```diff
@@ -302,4 +286,24 @@ public int getWascId() {
...
+        return newAlert()
+                .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                .setAttack(attack)
+                .setParam(param)
+                .setMessage(msg)
+                .setOtherInfo(getError(errorChar));
+    }
+
+    public List<Alert> getExampleAlerts() {
+        List<Alert> alerts = new ArrayList<>();
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java:302]: Use `List.of`.

**Comment by GRVial** [comment]
Changed all you pointed, will see if it builds now and then i will squash the commits.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java [hunk 3]
```diff
@@ -302,4 +285,26 @@ public int getWascId() {
...
+
+    @Override
+    public List<Alert> getExampleAlerts() {
+        List<Alert> alerts =
+                List.of(
+                        createAlert("example1%n%s...%n%s\n", "Example param", null, '1').build(),
+                        createAlert("example2%n%s...%n%s\n", "Example param", null, '2').build(),
+                        createAlert("example3%n%s...%n%s\n", "Example param", null, '3').build());
+
+        return alerts;
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java:308]: Can just return List.of without the intermediate storage.

File: addOns/ascanrules/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
+- The Format String Error scan rule now includes example alert functionality for documentation generation purposes (Issue 6119).
```
Comment (thc202) [inline addOns/ascanrules/CHANGELOG.md:7]: Should still be added under e.g. `### Changed`.

Comment (kingthorin) [inline addOns/ascanrules/CHANGELOG.md:7]:
```suggestion
### Added
- The Format String Error scan rule now includes example alert functionality for documentation generation purposes (Issue 6119).

```

**Review by kingthorin — APPROVED** [review]

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java [hunk 5]
```diff
@@ -302,4 +285,26 @@ public int getWascId() {
...
+                .setAttack(attack)
+                .setParam(param)
+                .setMessage(msg)
+                .setOtherInfo(getError(errorChar));
+    }
+
+    @Override
+    public List<Alert> getExampleAlerts() {
+        List<Alert> alerts =
+                List.of(
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java:303]: Return immediately, no need for the variable.

**Comment by kingthorin** [comment]
Crap, you dunno why that came up approved. Obviously I bungled something.

**Review by kingthorin — COMMENTED** [review]
Per previous comments 😉

**Comment by GRVial** [comment]
I'll look into it, seems like i forgot a sign off aswell.

**Comment by kingthorin** [comment]
@GRVial are you going to finish this?

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java [hunk 6]
```diff
@@ -302,4 +285,26 @@ public int getWascId() {
...
+                        createAlert("example1%n%s...%n%s\n", "Example param", null, '1').build(),
+                        createAlert("example2%n%s...%n%s\n", "Example param", null, '2').build(),
+                        createAlert("example3%n%s...%n%s\n", "Example param", null, '3').build());
```
[the remark this answers is not in the document]
Comment (GRVial) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/FormatStringScanRule.java:306]:
```suggestion
                        createAlert("%n%s%n%s%n%s%n%s\n", "param", null, '1').build(),
                        createAlert("%s%s%s%s%s%s", "param", null, '2').build(),
                        createAlert("%p %p %p %p %p %p", "param", null, '3').build());
```

So, i found some examples to use as the attack payload, but i'm not sure about what the param should be.

**Comment by GRVial** [comment]
> @GRVial are you going to finish this?

Yes, sorry for the delay. I'm having doubts about the relevant details on my examples.

**Comment by kingthorin** [comment]
Oops?

**Comment by GRVial** [comment]
Seems i broke something when squashing, was going to close and re-open the PR to rescope it manually but i cant, should i just open another PR? @kingthorin

**Comment by GRVial** [comment]
Reopenning

**Comment by kingthorin** [comment]
Most of your changes or still here. Just remove you local branch. Create one with the same name. Re-implement the changes. Push back to origin (your fork … you’ll probably need to force push) and re-open this.

**Comment by GRVial** [comment]
I still can't re-open. Probably doing something wrong... I'll try some more later.

**Comment by kingthorin** [comment]
You won't be able to re-open until after the branch is re-pushed with the appropriate changes.

**Comment by thc202** [comment]
This no longer seems to be in sync with the repo, the PR shows 2 commits while the `main` in the repo shows 1.

**Comment by kingthorin** [comment]
Hmmm maybe it will require a new PR then, I didn't know that could happen.

**Comment by thc202** [comment]
I guess the default branch has more restrictions/conditions, they could try push `8574f6d6ed28a9dbc042e3feed962dee5c9f3550` which is the commit this PR has and see if it allows to reopen.

**Comment by GRVial** [comment]
> I guess the default branch has more restrictions/conditions, they could try push `8574f6d6ed28a9dbc042e3feed962dee5c9f3550` which is the commit this PR has and see if it allows to reopen.

Worked, nice. Thanks.

**Review by thc202 — APPROVED** [review]

**Review by psiinon — APPROVED** [review]

**Pull request merged**